### ISSUE TITLE
Compilation Performance when adding Custom Tags

### ISSUE THREAD

**Issue by atonse** [edgurgel/solid#131]
Thank you for this awesome library, we love it 😄 

We've noticed that when we add custom tags (of which we've added 8) it slows down our compiles tremendously when processing those tags. So just the tags take about 20 seconds to compile.

Do you know if this is this something that's just a limitation of `nimble_parsec` and some kind of Big(O) issue with the underlying algorithms or potentially how we're constructing our tags that's resulting in way more parser combinations than needed?

Here's an example of a custom tag:

In Liquid it is:
```liquid
{% image file_id %}
```

The elixir:
```elixir
  @impl true
  def spec(_parser) do
    space = Literal.whitespace(min: 0)

    ignore(BaseTag.opening_tag())
    |> ignore(space)
    |> ignore(string("image"))
    |> ignore(space)
    |> tag(Argument.argument(), :file_id)
    |> ignore(space)
    |> ignore(BaseTag.closing_tag())
  end
```

**Comment by edgurgel** [comment]
Yeah the parser is quite complex and to keep it fast we avoid using lots of "defcombinators". But it's something we can revisit once we have a proper set of benchmarks :thinking:.

**Comment by atonse** [comment]
Thank you again for your work on this.

Looks like it was a huge amount of refactoring. We'll review and incorporate it.

**Comment by edgurgel** [comment]
Thanks! Any feedback around the new parsing when building custom tags is welcome!

**Comment by edgurgel** [comment]
Hmm  this is tricky as Liquid has two main points of extension: Custom Tags and Custom Filters. What you shared is sort of a Custom Object which is not valid Liquid. 

Short answer is: No. The new parser is a bit more restrictive so that we could get the `{% liquid %}` tag for free for all tags including custom tags.

Long answer: 

I think there are two main options: 

* Migrate your existing templates using a regex to use a new custom tag to be like this: `{% secure_query_params email: user.email, id: user.id, example: 'hello' %}` . Then it should be pretty easy to migrate to 1.0.0 as this custom tag should be pretty easy to write. 
* Fork Solid and change the [Object](https://github.com/edgurgel/solid/blob/main/lib/solid/object.ex) to understand this custom object.

**Comment by edgurgel** [comment]
Great! In fact I might add a similar custom tag as an example to the tests as it's a good real world example.

**Comment by edgurgel** [comment]
Oh i think it's pretty close but it can be simpler! Argument.parse already gives you back the rest of the tokens `{:ok, %Argument{...}, rest}` once it has found an argument so your `parse_kv_pairs` can simple be:

```elixir
  defp parse_kv_pairs(tokens, acc \\ []) do
    case tokens do
      [{:identifier, _, key}, {:colon, _} | rest] ->
        with {:ok, value, rest} <- Solid.Argument.parse(rest) do
          acc = [{String.to_atom(key), value} | acc]

          case rest do
            [{:comma, _} | tokens_rest] -> parse_kv_pairs(tokens_rest, acc)
            # The list of tokens ALWAYS end with `{:end, meta}` so no need to match on empty list.
            [{:end, _}] -> {:ok, Enum.reverse(acc), rest}
           
            _ -> {:error, "Unexpected token sequence", Solid.Parser.meta_head(rest)}
          end
        end

      _ ->
        {:error, "Unexpected structure. Expected key: value", Solid.Parser.meta_head(tokens)}
    end
  end
```

Somewhat similar to what I've done here on RenderTag: 

https://github.com/edgurgel/solid/blob/4feb69d601a6af7769e3a020ffc08e9fda1c968a/lib/solid/tags/render_tag.ex#L53-L66

Does this make sense?

As a side-note I don't think you **need** that `String.to_atom(key)` as it can grow indefinitely the atom table causing memory issues. It doesn't look like you need it to be an atom.
